protocols.go

```diff
@@ -45,21 +46,76 @@ const (
-	Protocol{P_UDT, 0, "udt", CodeToVarint(P_UDT), false, nil},
-	Protocol{P_QUIC, 0, "quic", CodeToVarint(P_QUIC), false, nil},
-	Protocol{P_HTTP, 0, "http", CodeToVarint(P_HTTP), false, nil},
-	Protocol{P_HTTPS, 0, "https", CodeToVarint(P_HTTPS), false, nil},
-	Protocol{P_IPFS, LengthPrefixedVarSize, "ipfs", CodeToVarint(P_IPFS), false, TranscoderIPFS},
-	Protocol{P_UNIX, LengthPrefixedVarSize, "unix", CodeToVarint(P_UNIX), true, TranscoderUnix},
+	protoIP4,
+	protoTCP,
+	protoUDP,
+	protoDCCP,
+	protoIP6,
+	protoSCTP,
+	protoONION,
+	protoUTP,
+	protoUDT,
+	protoQUIC,
+	protoHTTP,
+	protoHTTPS,
+	protoP2P,
+	protoUNIX,
+}
+
+var (
+	protoIP4  = Protocol{P_IP4, 32, "ip4", CodeToVarint(P_IP4), false, TranscoderIP4}
+	protoTCP  = Protocol{P_TCP, 16, "tcp", CodeToVarint(P_TCP), false, TranscoderPort}
+	protoUDP  = Protocol{P_UDP, 16, "udp", CodeToVarint(P_UDP), false, TranscoderPort}
+	protoDCCP = Protocol{P_DCCP, 16, "dccp", CodeToVarint(P_DCCP), false, TranscoderPort}
+	protoIP6  = Protocol{P_IP6, 128, "ip6", CodeToVarint(P_IP6), false, TranscoderIP6}
+	// these require varint
+	protoSCTP  = Protocol{P_SCTP, 16, "sctp", CodeToVarint(P_SCTP), false, TranscoderPort}
+	protoONION = Protocol{P_ONION, 96, "onion", CodeToVarint(P_ONION), false, TranscoderOnion}
+	protoUTP   = Protocol{P_UTP, 0, "utp", CodeToVarint(P_UTP), false, nil}
+	protoUDT   = Protocol{P_UDT, 0, "udt", CodeToVarint(P_UDT), false, nil}
+	protoQUIC  = Protocol{P_QUIC, 0, "quic", CodeToVarint(P_QUIC), false, nil}
+	protoHTTP  = Protocol{P_HTTP, 0, "http", CodeToVarint(P_HTTP), false, nil}
+	protoHTTPS = Protocol{P_HTTPS, 0, "https", CodeToVarint(P_HTTPS), false, nil}
+	protoP2P   = Protocol{P_P2P, LengthPrefixedVarSize, "ipfs", CodeToVarint(P_P2P), false, TranscoderP2P}
+	protoUNIX  = Protocol{P_UNIX, LengthPrefixedVarSize, "unix", CodeToVarint(P_UNIX), true, TranscoderUnix}
+)
+
+var ProtocolsByName = map[string]Protocol{}
+
+func init() {
+	for _, p := range Protocols {
+		ProtocolsByName[p.Name] = p
+	}
+
+	// explicitly set both of these
+	ProtocolsByName["p2p"] = protoP2P
+	ProtocolsByName["ipfs"] = protoP2P
+}
+
+// SwapToP2pMultiaddrs is a function to make the transition from /ipfs/...
+// multiaddrs to /p2p/... multiaddrs easier
+// The first stage of the rollout is to ship this package to all users so
+// that all users of multiaddr can parse both /ipfs/ and /p2p/ multiaddrs
+// as the same code (P_P2P). During this stage of the rollout, all addresses
+// with P_P2P will continue printing as /ipfs/, so that older clients without
+// the new parsing code won't break.
+// Once the network has adopted the new parsing code broadly enough, users of
+// multiaddr can add a call to this method to an init function in their codebase.
+// This will cause any P_P2P multiaddr to print out as /p2p/ instead of /ipfs/.
+// Note that the binary serialization of this multiaddr does not change at any
+// point. This means that this code is not a breaking network change at any point
```

Comment:
I think I've missed a memo about this. We should make sure we're doing these stages of rollout at the same time in JS-land, correct?

Reply:
This stage can go forward at any time (the sooner the better). However, we'll need to wait for JS-land before moving on to the second stage.